ulam-spiral/src/main.rs

```diff
@@ -14,8 +14,90 @@
 // limitations under the License.
 //
 
+use std::ops::{Index, IndexMut};
 use image::{ImageBuffer, Luma};
 
+struct SquareMatrix {
+    elems: Vec<usize>
+}
+
+impl SquareMatrix {
+    pub fn new(size: usize) -> Self {
+        assert!(size > 0 && size % 2 != 0, "Matrix size must be odd and bigger than 0!");
+
+        let mut elems = vec![0; size * size];
+
+        let mut dir: Directions = Directions::RIGHT;
+        let mut step: usize = 1;
+        let mut xy_cursor: (usize, usize) = (size / 2, size / 2);
+        elems[(xy_cursor.1 * size) + xy_cursor.0] = 1;
+
+        let mut i = 2;
+        'outer: loop {
+            for _ in 0..2 { // Step is incremented every two directions
+                for _ in 0..step { // Number of steps before turning direction
+                    if i > (size * size) {
+                        break 'outer;
+                    }
+
+                    match dir {
+                        Directions::RIGHT => xy_cursor.0 += 1,
+                        Directions::UP => xy_cursor.1 -= 1,
+                        Directions::LEFT => xy_cursor.0 -= 1,
+                        Directions::DOWN => xy_cursor.1 += 1,
+                    }
+
+                    elems[(xy_cursor.1 * size) + xy_cursor.0] = i;
+                    i = i + 1;
+                }
+                dir = dir.rotate_counter_clockwise();
+            }
+            step = step + 1;
+        }
+
+        Self {
+            elems
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        return self.elems.len();
+    }
+}
+
+impl Index<usize> for SquareMatrix {
+    type Output = usize;
+
+    fn index(&self, idx: usize) -> &usize {
+        return &self.elems[idx];
+    }
+}
+
+impl IndexMut<usize> for SquareMatrix {
+    fn index_mut(&mut self, idx: usize) -> &mut usize {
+        return &mut self.elems[idx];
+    }
+}
+
+enum Directions {
+    RIGHT = 0,
```

Comment:
no need to specify values, also we typically format enum variants with `PascalCase`